crates/uv-toolchain/src/find.rs

```diff
@@ -13,3 +18,89 @@ pub static TOOLCHAIN_DIRECTORY: Lazy<PathBuf> = Lazy::new(|| {
         PathBuf::from,
     )
 });
+
+/// An installed Python toolchain.
+#[derive(Debug, Clone)]
+pub struct Toolchain {
+    path: PathBuf,
+}
+
+impl Toolchain {
+    pub fn executable(&self) -> PathBuf {
+        if cfg!(windows) {
+            self.path.join("install").join("python.exe")
+        } else if cfg!(unix) {
+            self.path.join("install").join("bin").join("python3")
+        } else {
+            unimplemented!("Only Windows and Unix systems are supported.")
+        }
+    }
+}
+
+/// Return the toolchains that satisfy the given Python version on this platform.
+///
+/// ## Errors
+///
+/// - The platform metadata cannot be read
+/// - A directory in the toolchain directory cannot be read
+pub fn toolchains_for_version(version: &PythonVersion) -> Result<Vec<Toolchain>, Error> {
+    let platform_key = platform_key_from_env()?;
+
+    // TODO(zanieb): Consider returning an iterator instead of a `Vec`
+    //               Note we need to collect paths regardless for sorting by version.
+
+    let toolchain_dirs = match fs_err::read_dir(TOOLCHAIN_DIRECTORY.to_path_buf()) {
+        Ok(toolchain_dirs) => {
+            // Collect sorted directory paths; `read_dir` is not stable across platforms
+            let directories: BTreeSet<_> = toolchain_dirs
+                .filter_map(|read_dir| match read_dir {
+                    Ok(entry) => match entry.file_type() {
+                        Ok(file_type) => file_type.is_dir().then_some(Ok(entry.path())),
+                        Err(err) => Some(Err(err)),
+                    },
+                    Err(err) => Some(Err(err)),
+                })
+                .collect::<Result<_, std::io::Error>>()
+                .map_err(|err| Error::ReadError {
+                    dir: TOOLCHAIN_DIRECTORY.to_path_buf(),
+                    err,
+                })?;
+            directories
+        }
+        Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+            return Ok(Vec::new());
+        }
+        Err(err) => {
+            return Err(Error::ReadError {
+                dir: TOOLCHAIN_DIRECTORY.to_path_buf(),
+                err,
+            })
+        }
+    };
+
+    Ok(toolchain_dirs
+        .into_iter()
+        // Sort "newer" versions of Python first
+        .rev()
+        .filter_map(|path| {
+            if path.file_name().is_some_and(|filename| {
+                filename
+                    .to_string_lossy()
```

Comment:
Maybe just do this once, instead of in both conditions? Since it _can_ allocate.